File: krypton/src/main/java/me/neznamy/tab/platforms/krypton/KryptonTabList.java
package me.neznamy.tab.platforms.krypton;

import lombok.RequiredArgsConstructor;
import me.neznamy.tab.shared.chat.IChatBaseComponent;
import me.neznamy.tab.shared.platform.TabList;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.kryptonmc.api.auth.GameProfile;
import org.kryptonmc.api.auth.ProfileProperty;
import org.kryptonmc.api.entity.player.TabListEntry;
import org.kryptonmc.api.world.GameMode;
import java.util.Collections;
import java.util.UUID;

@RequiredArgsConstructor
public class KryptonTabList implements TabList {

    private final KryptonTabPlayer player;

    @Override
    public void removeEntry(@NotNull UUID entryId) {
        player.getPlayer().getTabList().removeEntry(entryId);
    }

    @Override
    public void updateDisplayName(@NotNull UUID entryId, @Nullable IChatBaseComponent displayName) {
        TabListEntry entry = player.getPlayer().getTabList().getEntry(entryId);
        if (entry != null) entry.setDisplayName(displayName == null ? null : displayName.toAdventureComponent());
    }

    @Override
    public void updateLatency(@NotNull UUID entryId, int latency) {
        TabListEntry entry = player.getPlayer().getTabList().getEntry(entryId);
        if (entry != null) entry.setLatency(latency);
    }

    @Override
    public void updateGameMode(@NotNull UUID entryId, int gameMode) {
        TabListEntry entry = player.getPlayer().getTabList().getEntry(entryId);
        if (entry != null) entry.setGameMode(GameMode.values()[gameMode]);
    }

    @Override
    public void addEntry(@NotNull Entry entry) {
        GameProfile profile = createGameProfile(entry.getUniqueId(), entry.getName(), entry.getSkin());
        player.getPlayer().getTabList().createEntryBuilder(entry.getUniqueId(), profile)
                .displayName(entry.getDisplayName() == null ? null : entry.getDisplayName().toAdventureComponent())
                .gameMode(GameMode.values()[entry.getGameMode()])
                .latency(entry.getLatency())
                .listed(true)
                .buildAndRegister();
    }

    @Override
    public void setPlayerListHeaderFooter(@NotNull IChatBaseComponent header, @NotNull IChatBaseComponent footer) {
        player.getPlayer().getTabList().setHeaderAndFooter(header.toAdventureComponent(), footer.toAdventureComponent());
    }

    private @NotNull GameProfile createGameProfile(@NotNull UUID uuid, @Nullable String name, @Nullable Skin skin) {
        String newName = name == null ? "" : name;
        if (skin == null) return GameProfile.of(newName, uuid);
        ProfileProperty property = ProfileProperty.of(TabList.TEXTURES_PROPERTY, skin.getValue(), skin.getSignature());
        return GameProfile.of(newName, uuid, Collections.singletonList(property));
    }
}

File: krypton/src/main/kotlin/me/neznamy/tab/platforms/krypton/KryptonTabList.kt
package me.neznamy.tab.platforms.krypton

import me.neznamy.tab.api.chat.IChatBaseComponent
import me.neznamy.tab.api.tablist.TabListEntry
import me.neznamy.tab.shared.tablist.BulkUpdateTabList
import net.kyori.adventure.text.serializer.gson.GsonComponentSerializer
import org.kryptonmc.api.auth.GameProfile
import org.kryptonmc.api.world.GameMode
import org.kryptonmc.krypton.packet.out.play.PacketOutPlayerInfoRemove
import org.kryptonmc.krypton.packet.out.play.PacketOutPlayerInfoUpdate
import java.util.*

class KryptonTabList(private val player: KryptonTabPlayer) : BulkUpdateTabList() {

    override fun removeEntries(entries: MutableCollection<UUID>) {
        player.sendPacket(PacketOutPlayerInfoRemove(ArrayList(entries)))
    }

    override fun updateDisplayNames(entries: MutableMap<UUID, IChatBaseComponent?>) {
        player.sendPacket(PacketOutPlayerInfoUpdate(EnumSet.of(PacketOutPlayerInfoUpdate.Action.UPDATE_DISPLAY_NAME), entries.map { entry ->
            PacketOutPlayerInfoUpdate.Entry(
                entry.key,
                GameProfile.of("", entry.key),
                false,
                0,
                GameMode.SURVIVAL,
                entry.value?.toString(player.version)?.let { GsonComponentSerializer.gson().deserialize(it) },
                null
            )
        }))
    }

    override fun updateLatencies(entries: MutableMap<UUID, Int>) {
        player.sendPacket(PacketOutPlayerInfoUpdate(EnumSet.of(PacketOutPlayerInfoUpdate.Action.UPDATE_LATENCY), entries.map { entry ->
            PacketOutPlayerInfoUpdate.Entry(
                entry.key,
                GameProfile.of("", entry.key),
                false,
                entry.value,
                GameMode.SURVIVAL,
                null,
                null
            )
        }))
    }

    override fun updateGameModes(entries: MutableMap<UUID, Int>) {
        player.sendPacket(PacketOutPlayerInfoUpdate(EnumSet.of(PacketOutPlayerInfoUpdate.Action.UPDATE_GAME_MODE), entries.map { entry ->
            PacketOutPlayerInfoUpdate.Entry(
                entry.key,
                GameProfile.of("", entry.key),
                false,
                0,
                GameMode.values()[entry.value],
                null,
                null
            )
        }))
    }

    override fun addEntries(entries: MutableCollection<TabListEntry>) {
        player.sendPacket(PacketOutPlayerInfoUpdate(EnumSet.of(PacketOutPlayerInfoUpdate.Action.ADD_PLAYER), entries.map { entry ->
            PacketOutPlayerInfoUpdate.Entry(
                entry.uniqueId,
                GameProfile.of(entry.name ?: "", entry.uniqueId),
                entry.isListed,
                entry.latency,
                GameMode.values()[entry.gameMode],
                entry.displayName?.toString(player.version)?.let { GsonComponentSerializer.gson().deserialize(it) },
                null //TODO chat session?
            )
        }))
    }
}